# Promise

> [출처1](https://ko.javascript.info/), [출처2](https://joshua1988.github.io/web-development/javascript/javascript-asynchronous-operation/)

## 용어 정리

- 비동기 처리
  - 특정 코드의 연산이 끝날 때까지 코드의 실행을 멈추지 않고 다음 코드를 먼저 실행하는 자바스크립트의 특성.
- callback
  1. 다른 함수의 인자로써 이용되는 함수.
  2. 어떤 이벤트에 의해 호출되어지는 함수.

- Promise
  - 비동기적으로 실행하는 작업의 결과(성공 or 실패)를 나타내는 객체.
  - 주목해야할 부분은 **객체**라는 점인데, 비동기의 결과를 객체화 시킨다는 것이 Promise의 가장 큰 장점이자 특징이다.
- Promise chaining
  - 우선 체이닝이란, 메서드가 객체를 반환하면 메서드의 반환 값인 객체를 통해 또 다른 함수를 호출하는 패턴을 말한다.
  - 프로미스 체이닝은 프로미스가 반환하는 값을 성공 시 then, 실패시 catch로 받아 잇는 패턴을 말함.
- async await
  - 자바스크립트 비동기 처리 패턴 중 하나.

<br/>

## 비동기 처리

가장 흔한 사례는 제이쿼리의 ajax. 화면에 표시할 이미지나 데이터를 서버에서 불러와 표시해야 함. ajax 통신으로 해당 데이터를 서버로부터 가져올 수 있다.

```js
function getData() {
  let tableData;
  $.get('https://domain.com/products/1', response => {
    tableData = response;
  });
  return tableData;
}

console.log(getData()); // undefined
```

왜 console에 undefined가 찍힐까? $.get으로 데이터를 요청하고 받아올 때까지 기다려주지 않고 `return tableData;` 를 실행했기 때문. 초기 값을 설정하지 않음 tableData의 undefined를 출력했기 때문이다.

**이렇게 특정 로직의 실행이 끝날 때까지 기다려주지 않고 나머지 코드를 먼저 실행하는 것이 비동기 처리**. 비동기 처리가 필요한 이유는 화면에서 서버로 데이터를 요청했을 때 서버가 언제 그 요청에 대한 응답을 줄지도 모르는데 마냥 다른 코드를 실행하지 않고 기다릴 수 없기 때문.

다른 사례는 `setTimeout()`. 코드를 바로 실행하지 않고 지정한 시간만큼 기다렸다가 로직을 실행함.

```js
console.log('Hello');

setTimeout(() => {
  console.log('Bye');
}, 3000);

console.log('Hello Again');
```

Hello -> Hello Again -> 3초 후 Bye. 비동기이기 때문에 기다리지 않고 실행해버림.

<br/>

## 콜백

비동기 처리 방식의 문제점을 해결하기 위해 콜백 함수를 이용해보자.

```js
function getData(callbackFunc) {
  $.get('https://domain.com/products/1', response => {
    callbackFunc(response);  // 서버에서 받은 데이터 response를 callbackFunc() 함수에 넘겨줌
  });
}

getData(tableData => {
  console.log(tableData);  // $.get()의 response 값이 tableData에 전달됨.
});
```

이렇게 콜백 함수를 사용하면 특정 로직이 끝났을 때 원하는 동작을 실행시킬 수 있음.

### 콜백 지옥

비동기 처리 로직을 위해 콜백 함수를 연속해서 사용할 때 발생하는 문제 중 하나 콜백 지옥.

```js
$.get('url', response => {
  parseValue(response, id => {
    auth(id, result => {
      display(result, text => {
        console.log(text);
      });
    });
  });
});
```

데이터를 받아와 화면에 표시하기까지 인코딩, 사용자 인증 등을 처리. 이 모든 과정을 비동기로 처리해야한다고 하면 위와 같이 콜백 안에 콜백을 계속 무는 형식으로 해야했음. 구조 난해.

먼저 코딩 패턴으로 해결해보자.

```js
function parseValueDone(id) {
  auth(id, authDone);
}
function authDone(result) {
  display(result, displayDone);
}
function displayDone(text) {
  console.log(text);
}
$.get('url', response => {
  parseValue(response, parseValueDone);
});
```

더 편하게 Promise를 사용할 수 있다.

<br/>

## Promise

위에서 봤던 코드를 Promise화 해보면,

```js
function getData(callback) {
  // new Promise() 추가
  return new Promise((resolve, reject) => {
    $.get('url', response => {
      resolve(response);
    });
  });
}

// getData()의 실행이 끝나면 호출되는 then()
getData().then(tableData => {
  // resolve() 의 결과 값이 여기로 전달됨.
  console.log(tableData);  // $.get()의 response 값이 tableData에 전달.
})
```

콜백의 구조에서 `new Promise()`, `resolve()`, `then()` 과 같은 프로미스 API를 사용한 구조로 바뀌었음.

### Promise의 3가지 상태

- Pending(대기): 비동기 처리 로직이 아직 완료되지 않은 상태
- Fulfilled(이행): 비동기 처리가 완료되어 결과 값을 반환해준 상태
- Rejected(실패): 비동기 처리가 실패하거나 오류가 발생한 상태

#### Pending(대기)

`new Promise()` 메서드를 호출하면 대기 상태가 됨. 콜백 함수를 선언할 수 있고, 콜백의 인자는 `resolve`, `reject` 임.

#### Fulfilled(이행)

`new Promise((resolve, reject) => resolve());` 성공. 이렇게 되면 `then()` 을 이용해 처리 결과 값을 받을 수 있다.

#### Rejected(실패)

`new Promise((resolve, reject) => reject());` 실패. `catch()` 를 이용해 실패 처리의 값을 받을 수 있다.

<br/>

## Promise 체이닝

여러 개의 프로미스를 연결해 사용할 수 있다. `then()` 메서드를 호출하고 나면 새로운 프로미스 객체가 반환된다.

```js
function getData() {
  return new Promise({
    (...)
  });
}

// then()으로 여러 개의 프로미스를 연결한 형식
getData()
  .then(data => {
    (...)
  })
  .then(data => {
    (...)
  })
  .then(data => {
    (...)
  });
```

### 체이닝 패턴의 장단점

- 장점: 코드량이 줄고, 코드가 좀 더 간결해져 하나의 문장처럼 읽히게 할 수 있음. 함수를 쪼개는 방법을 생각하게 되고, 혼자서 너무 많은 일을 처리하려는 함수보다 좀 더 작고 특화된 함수를 만들 수 있게 된다는 점.
- 단점: 디버깅하기 어려움. 어느 라인에서 에러가 발생했는지 알아내더라도, 그 라인에서 수행하는 일이 너무 많을 수 있다.

### 프로미스의 에러 처리 방법

실제 서비스에는 네트워크 연결, 서버 문제 등으로 인해 오류가 발생할 수 있다. 따라서, 프로미스의 에러 처리 방법도 알고 있어야 함. 2가지 방법이 있다.

#### 1. then()의 두 번째 인자로 에러 처리

```js
getData().then(handleSuccess, handleError);
```

#### 2. catch()를 이용

```js
getData().then().catch();
```

2가지 방법 모두 프로미스의 `reject()` 메서드가 호출되어 실패 상태가 된 경우 실행됨.

가급적 2번 catch를 사용하자. 왜냐?

```js
// then()의 두 번째 인자로는 감지하지 못하는 오류
function getData() {
  return new Promise((resolve, reject) => {
    resolve('hi');
  });
}

getData().then(result => {
  console.log(result);
  throw new Error("Error in then()"); // Uncaught (in promise) Error: Error in then()
}, err => {
  console.log('then error : ', err);
});
```

getData() 함수의 프로미스에서 resolve() 메서드를 호출해 정상적으로 처리했지만, then() 의 **첫 번째 콜백 함수 내부에서 오류가 나는 경우 오류를 제대로 잡아내지 못**한다. 하지만 catch로 처리하면 잘 잡아냄.

<br/>

## Async await

```js
async function 함수명() {
  await 비동기_처리_메서드_명();
}
```

함수의 앞에 async라는 예약어를 붙이고, 비동기 처리 코드 앞에 await를 붙인다. 주의점은 비동기 처리 메서드가 반드시 프로미스 객체를 반환해야 await가 의도한 대로 동작한다. Promise 체이닝 보다 비동기 코드의 겉모습을 더 깔끔하고 읽기 쉽게 한다.

### Async await 예외처리

try catch를 이용하면 된다.

```js
async function 함수명() {
  try {
    await 비동기_처리_메서드_명(); 
  } catch (error) {
    (...)  // 에러처리 로직
  }
}
```
